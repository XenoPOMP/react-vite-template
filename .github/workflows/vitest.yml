name: 'Testing'
on:
  push:
  pull_request:

jobs:
  test:
    runs-on: ubuntu-latest

    permissions:
      # Required to checkout the code
      contents: read
      # Required to put a comment into the pull-request
      pull-requests: write

    steps:
      - uses: actions/checkout@v2
      - name: 'Install Node'
        uses: actions/setup-node@v2
        with:
          node-version: '16.x'
      - name: 'Install Deps'
        run: npm i yarn -g && yarn
      - name: 'Test'
        run: yarn coverage
      - name: 'Report Coverage'
        if: always() # Also generate the report if tests are failing
        uses: davelosert/vitest-coverage-report-action@v2
