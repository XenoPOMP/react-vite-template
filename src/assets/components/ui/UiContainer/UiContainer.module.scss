@import "src/assets/styles/templates";

.container {
  @extend %uiContainer;
}