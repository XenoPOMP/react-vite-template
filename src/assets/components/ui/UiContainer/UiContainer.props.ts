import { CSSProperties } from 'react';

export interface UiContainerProps {
	className?: string;
	style?: CSSProperties;
}
