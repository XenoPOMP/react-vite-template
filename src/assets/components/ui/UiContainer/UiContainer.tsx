import { PropsWith } from '@xenopomp/advanced-types';

import cn from 'classnames';
import { FC } from 'react';

import styles from './UiContainer.module.scss';
import type { UiContainerProps } from './UiContainer.props';

interface IUiContainerNestedProps
  extends PropsWith<'children' | 'id' | 'className', UiContainerProps> {}

const UiContainer: FC<IUiContainerNestedProps> & {
  Section: FC<IUiContainerNestedProps>;
  Article: FC<IUiContainerNestedProps>;
  Header: FC<IUiContainerNestedProps>;
} = ({ children, className, id }) => {
  return (
    <div id={id} className={cn(styles.container, className)}>
      {children}
    </div>
  );
};

UiContainer.Section = ({ children, className, id }) => (
  <section id={id} className={cn(styles.container, className)}>
    {children}
  </section>
);
UiContainer.Article = ({ children, className, id }) => (
  <article id={id} className={cn(styles.container, className)}>
    {children}
  </article>
);
UiContainer.Header = ({ children, className, id }) => (
  <header id={id} className={cn(styles.container, className)}>
    {children}
  </header>
);

export default UiContainer;
