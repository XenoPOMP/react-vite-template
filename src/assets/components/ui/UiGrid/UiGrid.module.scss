@import "src/assets/styles/templates";

.grid {
  @extend %uiGrid;
}