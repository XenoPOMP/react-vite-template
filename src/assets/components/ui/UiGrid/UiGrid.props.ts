import { GridSettings } from '@utils/getUiSx';

export interface UiGridProps extends GridSettings {}
