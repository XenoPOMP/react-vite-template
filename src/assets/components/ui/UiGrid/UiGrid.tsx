import { PropsWith } from '@xenopomp/advanced-types';

import cn from 'classnames';
import { FC } from 'react';

import getUiSx from '@utils/getUiSx';

import styles from './UiGrid.module.scss';
import type { UiGridProps } from './UiGrid.props';

interface IUiGridNestedProps
  extends PropsWith<'children' | 'id' | 'className', UiGridProps> {}

const UiGrid: FC<IUiGridNestedProps> & {
  Section: FC<IUiGridNestedProps>;
  Article: FC<IUiGridNestedProps>;
} = ({ children, className, id, cols, rows, gap, maxWidth }) => {
  return (
    <div
      style={{
        ...getUiSx({
          cols,
          rows,
          gap,
          maxWidth,
        }),
      }}
      className={cn(styles.grid, className)}
      id={id}
    >
      {children}
    </div>
  );
};

UiGrid.Section = ({ children, id, className, cols, rows, gap, maxWidth }) => (
  <section
    style={{
      ...getUiSx({
        cols,
        rows,
        gap,
        maxWidth,
      }),
    }}
    id={id}
    className={cn(styles.grid, className)}
  >
    {children}
  </section>
);

UiGrid.Article = ({ children, id, className, cols, rows, gap, maxWidth }) => (
  <article
    style={{
      ...getUiSx({
        cols,
        rows,
        gap,
        maxWidth,
      }),
    }}
    id={id}
    className={cn(styles.grid, className)}
  >
    {children}
  </article>
);

export default UiGrid;
