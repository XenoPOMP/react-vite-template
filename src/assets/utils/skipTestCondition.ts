import useEnv, { TestingMode } from '@hooks/useEnv';

const { TESTING_MODE } = useEnv();

/**
 * Determines skip test condition.
 * @param {TestingMode} mode       target mode
 * @returns {boolean}              false if current mode isn`t the same as target one
 *                                 or current mode is FULLSTACK
 */
const skipTestCondition = (mode: TestingMode): boolean => {
  if (TESTING_MODE === 'FULLSTACK') {
    return false;
  }

  return TESTING_MODE !== mode;
};

export { skipTestCondition as default };
