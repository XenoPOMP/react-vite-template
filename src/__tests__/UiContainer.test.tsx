import { render } from '@testing-library/react';
import { describe, expect, test } from 'vitest';

import UiContainer from '@ui/UiContainer/UiContainer';
import UiGrid from '@ui/UiGrid/UiGrid';

import skipTestCondition from '@utils/skipTestCondition';

describe.skipIf(skipTestCondition('FRONTEND'))('UI container', () => {
  test('Match snapshot', () => {
    expect(
      render(
        <>
          <UiContainer>Container</UiContainer>
          <UiContainer.Header>Container</UiContainer.Header>
          <UiContainer.Section>Container</UiContainer.Section>
          <UiContainer.Article>Container</UiContainer.Article>
        </>,
      ),
    ).toMatchSnapshot();

    expect(
      render(
        <>
          <UiGrid>Container</UiGrid>
          <UiGrid.Section>Container</UiGrid.Section>
          <UiGrid.Article>Container</UiGrid.Article>
        </>,
      ),
    ).toMatchSnapshot();
  });
});
