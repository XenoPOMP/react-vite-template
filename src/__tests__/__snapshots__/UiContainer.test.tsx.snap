// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UI container > Match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_container_0c0e90"
      >
        Container
      </div>
      <header
        class="_container_0c0e90"
      >
        Container
      </header>
      <section
        class="_container_0c0e90"
      >
        Container
      </section>
      <article
        class="_container_0c0e90"
      >
        Container
      </article>
    </div>
  </body>,
  "container": <div>
    <div
      class="_container_0c0e90"
    >
      Container
    </div>
    <header
      class="_container_0c0e90"
    >
      Container
    </header>
    <section
      class="_container_0c0e90"
    >
      Container
    </section>
    <article
      class="_container_0c0e90"
    >
      Container
    </article>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`UI container > Match snapshot 2`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_container_0c0e90"
      >
        Container
      </div>
      <header
        class="_container_0c0e90"
      >
        Container
      </header>
      <section
        class="_container_0c0e90"
      >
        Container
      </section>
      <article
        class="_container_0c0e90"
      >
        Container
      </article>
    </div>
    <div>
      <div
        class="_grid_8da033"
        style="--cols: 12; --rows: 1; --max-width: 1680px;"
      >
        Container
      </div>
      <section
        class="_grid_8da033"
        style="--cols: 12; --rows: 1; --max-width: 1680px;"
      >
        Container
      </section>
      <article
        class="_grid_8da033"
        style="--cols: 12; --rows: 1; --max-width: 1680px;"
      >
        Container
      </article>
    </div>
  </body>,
  "container": <div>
    <div
      class="_grid_8da033"
      style="--cols: 12; --rows: 1; --max-width: 1680px;"
    >
      Container
    </div>
    <section
      class="_grid_8da033"
      style="--cols: 12; --rows: 1; --max-width: 1680px;"
    >
      Container
    </section>
    <article
      class="_grid_8da033"
      style="--cols: 12; --rows: 1; --max-width: 1680px;"
    >
      Container
    </article>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
